lib/js/jsfetch.nim

```diff
@@ -0,0 +1,78 @@
+## - Fetch for the JavaScript target: https://developer.mozilla.org/docs/Web/API/Fetch_API
+from asyncjs import PromiseJs
+from httpcore import HttpMethod
+export PromiseJs, HttpMethod
+
+when not defined(js) and not defined(nimdoc):
+  {.fatal: "Module jsfetch is designed to be used with the JavaScript backend.".}
+when defined(nodejs):
+  {.warning: "By design 'fetch()' is defined on the web browser, but may not be on NodeJS.".}
+
+template method2cstring(metod: HttpMethod): cstring =
+  ## Template that takes an `HttpMethod` and returns an *Uppercase* `cstring`,
+  ## but *only* for the HTTP Methods that are supported by the fetch API.
+  ## High performance and minimal code compared to just `$(HttpMethod)`.
+  assert metod notin {HttpTrace, HttpOptions, HttpConnect}, "HTTP Method not supported by fetch API"
+  case metod
+  of HttpHead:   "HEAD".cstring
+  of HttpGet:    "GET".cstring
+  of HttpPost:   "POST".cstring
+  of HttpPut:    "PUT".cstring
+  of HttpDelete: "DELETE".cstring
+  of HttpPatch:  "PATCH".cstring
+  else:          "GET".cstring
+
+func fetch*(url: cstring): PromiseJs {.importcpp: "fetch(#)".}
+  ## `fetch()` API, Simple GET only (generates `fetch(url)`).
+  ##
+  ## .. code-block:: nim
+  ##   doAssert fetch("https://nim-lang.org") is PromiseJs
+
+func fetch*(url: cstring, httpMethod: static[HttpMethod], body: cstring): PromiseJs {.asmNoStackFrame.} =
+  ## `fetch()` API, for any HTTP Method (static overload for optimization).
+  ##
+  ## .. code-block:: nim
+  ##   doAssert fetch("https://my.api.org", HttpPost, "body") is PromiseJs
+  # See the generated code if you dont understand why this overload exists.
+  assert url.len > 0, "url must not be empty string"
+  const x: cstring = method2cstring(httpMethod)
+  {.emit: """return fetch(`url`, {method: "`x`", body: `body`});""".}
+
+func fetch*(url: cstring, httpMethod: HttpMethod, body: cstring): PromiseJs {.asmNoStackFrame.} =
+  ## .. code-block:: nim
+  ##   let mthd = HttpPost
+  ##   doAssert fetch("https://my.api.org", mthd, "body") is PromiseJs
+  # See the generated code if you dont understand why this overload exists.
+  assert url.len > 0, "url must not be empty string"
+  let x: cstring = method2cstring(httpMethod)
+  {.emit: """return fetch(`url`, {method: `x`, body: `body`});""".}
+
+func fetch*(url: cstring, httpMethod: HttpMethod, body, mode, cache, redirect,
```

Comment:
too many arguments; these be grouped inside an object or ref object?
otherwise this'll results in verbose code that just passes arguments around